### ISSUE TITLE
How to access Lua globals by NeoVim

### ISSUE THREAD

**Issue by weilbith** [noib3/nvim-oxi#41]
Hey 👋🏾 

First of all, thank you very much for all the effort and time you put into this project. I really appreciate it and it gave the necessary motivation to start working on NeoVim plugins again.

I'm now in need to make some calls to functions that are globally defined in Lua by NeoVim. Means I want to access `vim.lsp`, `vim.diagnostic`, ... Unfortunately I don't see how this should be possible right now. But maybe I'm blind. 🙈 
I took a look into your noib3/nvim-compleet project to checkout how you do it there. But it seems like this includes a huge amount of boilerplate code with a bridge etc. pp. completely independent of this library here. I'm not sure if there is any better way to do it, because I honestly have no clue about this topic. But in any case, do you plan to integrate this feature into `nvim-oxi` too? It looks like you did already all the work for it in the other project, so it would be just a matter of putting it here too and wrap it nicely. Am I wrong?

Looking forward to hear what you think. 🙃

**Comment by noib3** [comment]
Hi, you're not blind, it's currently not possible to access the `vim.lsp` or `vim.diagnostic` tables, but they are on my todo list.

The whole bridge machinery I'm using in `nvim-compleet` is orthogonal to this problem and won't be necessary here. I guess `nvim_oxi::diagnostic` could already be implemented without many issues, whereas for `nvim_oxi::lsp` I'd like to have implemented #38 to expose `async` functions instead of callbacks like in the lua api ([`:h lsp-handler`](https://neovim.io/doc/user/lsp.html#lsp-handler)).

**Comment by weilbith** [comment]
Thanks for your fast response. 🙏🏾 
That sounds like good news. What is the best way to stay updated of such changes? Do you do releases with a changelog or should I subscribe to PRs?

Might be related:
In case I want to access not a global object, but anything else available in NeoVims Lua environment, how would I do that? So for example if I want to access `coc.nvim` its functions. Is this technically possible too?

**Comment by noib3** [comment]
> What is the best way to stay updated of such changes? Do you do releases with a changelog or should I subscribe to PRs?

The project follows semantic versioning and there's a [`CHANGELOG`](https://github.com/noib3/nvim-oxi/blob/master/CHANGELOG.md).

> So for example if I want to access coc.nvim its functions. Is this technically possible too?

Technically speaking there's really no difference between e.g. `vim.lsp` and any other global Lua object exposed by other plugins.

However the `vim.*` tables are built in the editor and they provide a stable API, so I'm much more inclined to have first class support for those. Supporting other arbitrary plugins from the community would become unmaintainable pretty quickly. If you really need to interact often with other Lua plugins which may or may not be available at runtime, a general purpose Rust-Lua crate like [mlua](https://github.com/khvzak/mlua) could be a better option.

**Comment by weilbith** [comment]
Yeah, I wouldn't expect any support for arbitrary plugins. Just an easy way to access them. But maybe I need to learn how `mlua` works and then I do it like that.

**Comment by ModProg** [comment]
Maybe there could be a way to have  procmacro you can use to define an api of a third party macro easily, I did build some here: https://github.com/ModProg/nvim-rust/blob/6edc2435c11e94699efdd574efd981dace69fc33/src/api.rs#L14 and if you would accept something like this in nvim-oxi, this could be exported and made available to a user that quickly wants to create an interface to another plugin where they just need to provide the necessary globals and functions+parameters.

**Comment by weilbith** [comment]
I'm sorry to abuse this issue, but could someone point me to how I could use nvim-oxi (with or without additional `mlua`) to "require" other modules of the NeoVim Lua environment from my Rust code? I'm totally fine to write my own interfaces for the plugins I intend to interact with and only these parts I need to. But currently I have no clue how to do this at all, I'm sorry.

**Comment by noib3** [comment]
The `Lua` object is a wrapper around the Lua state which is not exposed in `nvim-oxi`. This means that you can't just call Lua code from `nvim-oxi` yourself, it has to be exposed by us. Can I ask what's your use case for `require`ing Lua modules from Rust?

**Comment by weilbith** [comment]
Okay.
My use-case are two. First, I want to interact with language servers and people often complain that they want to use in-built LSP support or CoC or vim-lsp or... So I intend to have a trait for the language client and have some implementations for it. And from there on have some typesafe Rust environment I can develop the core/business-logic of my plugin. 
On the other "end" of the plugin I need to visually show some stuff to the user. Therefore I don't want to re-invent the wheel and allow the user to choose their UI plugin like Telescope, vim-fzf, etc. pp. Again, an abstract trait and some implementations of these.
I could write only the core in Rust and the rest in Lua as "glue-code". But I would really love to make it exclusively in Rust. Lua is simply not a serious language for me except for hacking around some smallish things. It is too error prone and fragile in my opinion. I have an almost complete rewrite of one of my plugins, trying to establish proper architecture, clean code, a lot of unit tests and so forth. But in the end it is still all super loose and doesn't feel great. Lua doesn't has many features and this bugs me.

Thanks for your support. 🙏🏾
